pkg/daemon/pinned_image_set.go

```diff
@@ -0,0 +1,1273 @@
+		klog.Errorf("Failed to get image set apply configs: %v", err)
+		return
+	}
+
+	err = upgrademonitor.UpdateMachineConfigNodeStatus(
+		&upgrademonitor.Condition{
+			State:   mcfgv1alpha1.MachineConfigNodePinnedImageSetsProgressing,
+			Reason:  "AsExpected",
+			Message: message,
+		},
+		nil,
+		metav1.ConditionFalse,
+		metav1.ConditionUnknown,
+		node,
+		p.mcfgClient,
+		applyCfg,
+		imageSets,
+		p.featureGatesAccessor,
+	)
+	if err != nil {
+		klog.Errorf("Failed to updated machine config node: %v", err)
+	}
+
+	// reset any degraded status
+	err = upgrademonitor.UpdateMachineConfigNodeStatus(
+		&upgrademonitor.Condition{
+			State:   mcfgv1alpha1.MachineConfigNodePinnedImageSetsDegraded,
+			Reason:  "AsExpected",
+			Message: "All is good",
+		},
+		nil,
+		metav1.ConditionFalse,
+		metav1.ConditionUnknown,
+		node,
+		p.mcfgClient,
+		applyCfg,
+		imageSets,
+		p.featureGatesAccessor,
+	)
+	if err != nil {
+		klog.Errorf("Failed to updated machine config node: %v", err)
+	}
+}
+
+func (p *PinnedImageSetManager) updateStatusError(pools []*mcfgv1.MachineConfigPool, statusErr error) {
+	node, err := p.nodeLister.Get(p.nodeName)
+	if err != nil {
+		klog.Errorf("Failed to get node %q: %v", p.nodeName, err)
+		return
+	}
+
+	applyCfg, imageSets, err := getImageSetApplyConfigs(p.imageSetLister, pools, nil)
+	if err != nil {
+		klog.Errorf("Failed to get image set apply configs: %v", err)
+		return
+	}
+
+	err = upgrademonitor.UpdateMachineConfigNodeStatus(
+		&upgrademonitor.Condition{
+			State:   mcfgv1alpha1.MachineConfigNodePinnedImageSetsProgressing,
+			Reason:  "ImagePrefetch",
+			Message: fmt.Sprintf("node is prefetching images: %s", node.Name),
+		},
+		nil,
+		metav1.ConditionTrue,
+		metav1.ConditionUnknown,
+		node,
+		p.mcfgClient,
+		applyCfg,
+		imageSets,
+		p.featureGatesAccessor,
+	)
+	if err != nil {
+		klog.Errorf("Failed to updated machine config node: %v", err)
+	}
+
+	applyCfg, imageSets, err = getImageSetApplyConfigs(p.imageSetLister, pools, statusErr)
+	if err != nil {
+		klog.Errorf("Failed to get image set apply configs: %v", err)
+		return
+	}
+
+	err = upgrademonitor.UpdateMachineConfigNodeStatus(
+		&upgrademonitor.Condition{
+			State:   mcfgv1alpha1.MachineConfigNodePinnedImageSetsDegraded,
+			Reason:  "PrefetchFailed",
+			Message: "Error: " + statusErr.Error(),
+		},
+		nil,
+		metav1.ConditionTrue,
+		metav1.ConditionUnknown,
+		node,
+		p.mcfgClient,
+		applyCfg,
+		imageSets,
+		p.featureGatesAccessor,
+	)
+	if err != nil {
+		klog.Errorf("Failed to updated machine config node: %v", err)
+	}
+}
+
+// getImageSetApplyConfigs populates image set generation details for status updates to MachineConfigNode.
+func getImageSetApplyConfigs(imageSetLister mcfglistersv1alpha1.PinnedImageSetLister, pools []*mcfgv1.MachineConfigPool, statusErr error) ([]*machineconfigurationalphav1.MachineConfigNodeStatusPinnedImageSetApplyConfiguration, []mcfgv1alpha1.MachineConfigNodeSpecPinnedImageSet, error) {
```

Comment:
Not sure I quite understand the name, but I guess it matches `MachineConfigNodeStatusPinnedImageSetApplyConfiguration`

This is basically metadata right?

Reply:
yeah its status information that is consumed by MCN I am not familiar with the naming either.

Reply:
will all some docs